### PR TITLE
feat: add `stylelint-plugin`

### PR DESCRIPTION
**Pull request by valerybugakov**
## Context

Moving [the `stylelint-plugin` from the main Sourcegraph repo](https://github.com/sourcegraph/sourcegraph/blob/main/client/stylelint-plugin-sourcegraph/package.json) to leverage logic used for codemods later. 
This PR also introduces:

1. Typescript support for custom Stylelint rules.
https://github.com/sourcegraph/codemod/blob/3c7e6e68639c00d03ce8dde66d8736eaa2ccc693/packages/stylelint-plugin/src/rules/filenames-match-regex/filenames-match-regex.ts#L14-L18
3. Infrastructure to test new rules.
https://github.com/sourcegraph/codemod/blob/3c7e6e68639c00d03ce8dde66d8736eaa2ccc693/packages/stylelint-plugin/src/rules/no-restricted-imports/__tests__/no-restricted-imports.test.ts#L3-L38
5. Updates the `stylelint` package to the version used in the main Sourcegraph repo.
6. The new `no-restricted-imports` rule that we can use to ban imports from packages that [we want to migrate from](https://sourcegraph.slack.com/archives/C89KCDK5J/p1651645166164439). E.g., `bootstrap`.
https://github.com/sourcegraph/codemod/blob/3c7e6e68639c00d03ce8dde66d8736eaa2ccc693/packages/stylelint-plugin/src/rules/no-restricted-imports/no-restricted-imports.ts#L11

## Integration plan

1. After merging this PR, the `stylelint-plugin` will be published to NPM.
2. Then, we can install it in the main Sourcegraph repo and drop the JS version defined there.

## Test plan

1. `cd packages/stylelint-plugin && yarn link` 
2. `cd main-sourcegraph-repo && yarn link "@sourcegraph/stylelint-plugin-sourcegraph"`

### CODE REVIEW

**Review by umpox — APPROVED** [review]
Nice, LGTM

Open Q: Should this repo still be called `codemod`? Feels like codemod is more of a subset of `code-enforcement` or `code-quality` now, WDYT?

File: packages/stylelint-plugin/package.json [hunk 1]
```diff
@@ -0,0 +1,52 @@
+{
+  "name": "@sourcegraph/stylelint-plugin-sourcegraph",
+  "version": "1.0.0",
+  "description": "Custom stylelint rules for the main Sourcegraph repo",
```
Comment (umpox) [inline packages/stylelint-plugin/package.json:4]:
```suggestion
  "description": "Custom stylelint rules for Sourcegraph",
```

Is there any reason we shouldn't keep rules here generic and anything specific config/overrides should live in the main Sourcegraph repo?

Kinda like how we have the generic [eslint-config repo](https://github.com/sourcegraph/eslint-config) and specific overrides in `eslintrc` files in the main repo

Comment (valerybugakov) [reply packages/stylelint-plugin/package.json:4]: Yeah, that's the intention — rules in this repo will be generic. Here, I was trying to convey that the primary consumer of these rules will be the main Sourcegraph repo. Happy to update the description 👍

**Comment by valerybugakov** [comment]
> Should this repo still be called codemod? Feels like codemod is more of a subset of code-enforcement or code-quality now, WDYT?

@umpox, what about Eunomia? 😃 
> In [Greek mythology](https://en.wikipedia.org/wiki/Greek_mythology), Eunomia ([Ancient Greek](https://en.wikipedia.org/wiki/Ancient_Greek_language): Εὐνομία) was a minor goddess of law and legislation (her name can be translated as "good order", "governance according to good laws")

Seriously speaking, I like the `code-quality` option. We would also need to update internal links in the `package.json` files.

**Comment by umpox** [comment]
Hah, I actually kinda like Eunomia!

If we ever implement "chaos engineering" we'd have to have a repo called [Dysnomia](https://en.wikipedia.org/wiki/Dysnomia_(deity)) 😁

**Comment by valerybugakov** [comment]
All-in [Greek mythology](https://en.wikipedia.org/wiki/Greek_mythology)! 🚀

**Pull request merged**